fix for Java 8 Map merge method conflict

colinsurprenant opened the pull request:
fixes logstash-plugins/logstash-filter-multiline#10

related to JRuby issue jruby/jruby#1249

colinsurprenant commented:
I'd like a bench to be done to see if the commented "Java" implementation is faster than the "Ruby" one.

ph commented:
Result of the benchmark, the java code performance is really spread out, check the deviation.


```
Calculating -------------------------------------
          merge ruby    25.367k i/100ms
          merge java    17.053k i/100ms
        merge java 2    17.379k i/100ms
-------------------------------------------------
          merge ruby    336.218k (± 5.1%) i/s -      1.700M
          merge java    206.158k (±25.9%) i/s -    937.915k
        merge java 2    207.211k (±25.6%) i/s -    973.224k

Comparison:
          merge ruby:   336217.7 i/s
        merge java 2:   207211.5 i/s - 1.62x slower
          merge java:   206158.3 i/s - 1.63x slower
```

**Environement**

```
java version "1.8.0_20"
Java(TM) SE Runtime Environment (build 1.8.0_20-b26)
Java HotSpot(TM) 64-Bit Server VM (build 25.20-b23, mixed mode)
```
```
jruby 1.7.17 (1.9.3p392) 2014-12-09 fafd1a7 on Java HotSpot(TM) 64-Bit Server VM 1.8.0_20-b26 +jit [darwin-x86_64]
```

Related code: https://gist.github.com/f1d55e792e9cb3805d82

ph commented:
@colinsurprenant we should go ahead and use the merge, the performance is more stable and faster and we known it will follow the ruby contract.

colinsurprenant commented:
TIL benchmark-ips :D

I reworked a bit the benchmark and added a new test and using ruby merge is still the fastest.

```
Calculating -------------------------------------
          merge ruby    50.711k i/100ms
          merge java    42.195k i/100ms
        merge java 2    28.418k i/100ms
        merge java 3    41.135k i/100ms
-------------------------------------------------
          merge ruby    675.004k (± 6.5%) i/s -     20.132M
          merge java    567.134k (± 7.9%) i/s -     16.920M
        merge java 2    408.569k (±29.3%) i/s -     10.884M
        merge java 3    547.006k (± 7.8%) i/s -     16.289M

Comparison:
          merge ruby:   675003.9 i/s
          merge java:   567133.8 i/s - 1.19x slower
        merge java 3:   547006.0 i/s - 1.23x slower
        merge java 2:   408568.6 i/s - 1.65x slower
```

the code is at https://gist.github.com/colinsurprenant/c556af117e07375313f7

purbon commented:
LGTM

jsvd commented on lib/logstash/java_integration.rb:
```diff
@@ -24,24 +24,39 @@ def self.===(other)
-# https://github.com/jruby/jruby/blob/master/core/src/main/ruby/jruby/java/java_ext/java.util.rb
-# https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/java/proxies/MapJavaProxy.java
-class Java::JavaUtil::LinkedHashMap
+map_mixin = lambda do
```
Shouldn't this be in `lib/logstash/patches`?

colinsurprenant replied:
well, good point, most/all of the java_integration stuff could be considered a patch... can we merge this and create a ticket to see if we should refactor into lib/logstash/patches?

jsvd commented on lib/logstash/java_integration.rb:
```diff
@@ -24,24 +24,39 @@ def self.===(other)
 end
 
+Java::JavaUtil::LinkedHashMap.module_exec(&map_mixin)
+Java::JavaUtil::HashMap.module_exec(&map_mixin)
```
Is the patch to HashMap also needed? If so, please create a shared_example for both classes so that we have tests on both (right now deleting line 58 still makes all tests pass)

colinsurprenant replied:
+1

ph commented:
LGTM

jsvd commented:
LGTM

colinsurprenant commented:
:shipit:
